lit_gpt/speed_monitor.py

```diff
@@ -345,19 +354,35 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
         self.speed_monitor.eval_end(eval_elapsed)
 
 
+def flops_per_param(config: Config, n_params: int) -> int:
+    flops_per_token = 2 * n_params  # each parameter is used for a MAC (2 FLOPS) per network operation
+    # this assumes that all samples have a fixed length equal to the block size
+    # which is most likely false during finetuning
+    flops_per_seq = flops_per_token * config.block_size
+    attn_flops_per_seq = config.n_layer * 2 * 2 * (config.n_embd * (config.block_size**2))
+    return flops_per_seq + attn_flops_per_seq
+
+
 def estimate_flops(model: GPT) -> int:
-    """Measures estimated FLOPs for MFU: https://arxiv.org/abs/2205.05198"""
+    """Measures estimated FLOPs for MFU.
+
+    Refs:
+        * https://ar5iv.labs.arxiv.org/html/2205.05198#A1
+        * https://ar5iv.labs.arxiv.org/html/2204.02311#A2
+    """
     # using all parameters for this is a naive over estimation because not all model parameters actually contribute to
     # this FLOP computation (e.g. embedding, norm). For this reason, the result will be higher by a fixed percentage
     # (~10%) compared to the measured FLOPs, making those lower but more realistic.
     # For a proper estimate, this needs a more fine-grained calculation as in Appendix A of the paper.
-    n_params = sum(p.numel() for p in model.parameters())
-    # credit: https://github.com/mosaicml/examples/blob/release/v0.0.4/examples/llm/throughput/README.md#mfu-and-hfu
```

Comment:
don't we need this reference anymore?

Reply:
At this point this version is our own, the formulas come from the linked references in the docstring